Userland/Libraries/LibAudio/Output.h

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright (c) 2023, Gregory Bertilson <zaggy1024@gmail.com>
+ *
+ * SPDX-License-Identifier: BSD-2-Clause
+ */
+
+#pragma once
+
+#include <AK/AtomicRefCounted.h>
+#include <AK/Queue.h>
+#include <AK/Time.h>
+#include <LibAudio/OutputImplementation.h>
+#include <LibAudio/SampleFormats.h>
+#include <LibThreading/ConditionVariable.h>
+#include <LibThreading/MutexProtected.h>
+#include <LibThreading/Thread.h>
+
+namespace Audio {
+
+enum class OutputState {
+    Playing,
+    Suspended,
+};
+
+class Output : public AtomicRefCounted<Output> {
+public:
+    using TaskCompletionCallback = Function<void(Duration)>;
+
+    // Creates a new audio Output class.
+    //
+    // The initial_output_state parameter determines whether it will begin playback immediately.
+    //
+    // The AudioDataRequestCallback will be called when the Output needs more audio data to fill
+    // its buffers and continue playback.
+    //
+    // NOTE: The outputs will retain the previously set volume when running on the PulseAudio
+    //       backend. To start at 100% volume, call set_volume() immediately after creation.
+    static ErrorOr<NonnullRefPtr<Output>> create(OutputState initial_output_state, u32 sample_rate, u8 channels, u32 target_latency_ms, AudioDataRequestCallback&&);
+
+    ~Output();
+
+    // Sets the callback function that will be fired when any asynchronous function encounters an error.
+    void set_error_callback(Function<void(Error)>);
+    // Sets the callback function that will be fired whenever server consumes more data than is made available
+    // by the data request callback. It will fire when either the data request runs too long, or the data request
+    // returns no data. If all the input data has been exhausted and this event fires, that means that playback
+    // has ended.
+    void set_underrun_callback(Function<void()>);
+
+    // Resume playback from the suspended state, requesting new data for audio buffers as soon as possible.
+    void resume(TaskCompletionCallback = nullptr);
+    // Completes playback of any buffered audio data and then suspends playback and buffering.
+    void drain_buffer_and_suspend(TaskCompletionCallback = nullptr);
+    // Drops any buffered audio data and then suspends playback and buffering. This can used be to stop playback
+    // as soon as possible instead of waiting for remaining audio to play.
+    void discard_buffer_and_suspend(TaskCompletionCallback = nullptr);
+
+    // Returns a monotonically-increasing number of samples played on the device measured as a time duration.
+    //
+    // This function should be able to run from any thread safely.
+    ErrorOr<Duration> total_time_played();
```

Comment:
Why does this return a Duration if it supposedly "returns a [...] number of samples"? The comment doesn't really make sense if this returns a time period.